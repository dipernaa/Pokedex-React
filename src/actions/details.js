import api from './utils/api';
import { createAsyncAction } from './utils/asyncUtils';
import { GET_POKEMON_DETAILS } from '../constants/actionTypes';

function getPokemonDetails(url) {
  return createAsyncAction(GET_POKEMON_DETAILS, () => api.getAsync(`/api/details?url=${url}`));
}

export default {
  getPokemonDetails
};
