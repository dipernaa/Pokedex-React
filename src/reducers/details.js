import { GET_POKEMON_DETAILS } from '../constants/actionTypes';

export const initialState = {
  results: null,
  isLoading: false
};

export default function(state = initialState, action) {
  console.log(state);
  console.log(action);
  switch (action.type) {
    case GET_POKEMON_DETAILS.request:
      return {
        ...state,
        isLoading: true
      };

    case GET_POKEMON_DETAILS.failure:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON_DETAILS.success:
      return {
        ...state,
        results: action.data,
        isLoading: false
      };

    default:
      return state;
  }
}
