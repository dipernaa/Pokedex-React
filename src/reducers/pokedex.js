import { GET_POKEMON } from '../constants/actionTypes';

const sortPokemon = (arr) => (
  arr.sort((first, second) => {
    if (first.name > second.name) {
      return 1;
    } else if (first.name < second.name) {
      return -1;
    }

    return 0;
  })
);

export const initialState = {
  count: 0,
  next: null,
  previous: null,
  results: [],
  isLoading: false
};

export default function(state = initialState, action) {
  switch (action.type) {
    case GET_POKEMON.request:
      return {
        ...state,
        isLoading: true
      };

    case GET_POKEMON.failure:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON.success:
      return {
        ...state,
        count: action.data.count,
        next: action.data.next,
        previous: action.data.previous,
        results: sortPokemon(action.data.results.map((item, index) => ({
          ...item,
          id: index + 1
        }))),
        isLoading: false
      };

    default:
      return state;
  }
}
