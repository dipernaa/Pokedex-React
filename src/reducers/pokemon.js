import {
  GET_POKEMON,
  GET_POKEMON_DETAILS
} from '../constants/actionTypes';

const sortPokemon = (arr) => (
  arr.sort((first, second) => {
    if (first.name > second.name) {
      return 1;
    } else if (first.name < second.name) {
      return -1;
    }

    return 0;
  })
);

export const initialState = {
  count: 0,
  next: null,
  previous: null,
  results: [],
  isLoading: false
};

export default function(state = initialState, action) {
  // console.log(state);
  // console.log(action);
  switch (action.type) {
    case GET_POKEMON.request:
    case GET_POKEMON_DETAILS.request:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON.failure:
    case GET_POKEMON_DETAILS.failure:
      return {
        ...state,
        isLoading: false
      };

    case GET_POKEMON.success:
      return {
        ...state,
        count: action.data.count,
        next: action.data.next,
        previous: action.data.previous,
        results: sortPokemon(action.data.results.map((item, index) => ({
          ...item,
          id: index + 1
        }))),
        isLoading: false
      };

    case GET_POKEMON_DETAILS.success:
      return {
        ...state,
        results: sortPokemon(state.results.map(item => {
          if (item.id === action.data.id) {
            return {
              ...item,
              details: action.data
            };
          }

          return {
            ...item
          };
        })),
        isLoading: false
      };

    default:
      return state;
  }
}
