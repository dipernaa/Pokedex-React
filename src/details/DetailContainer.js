import React, { Component } from 'react';
import { connect } from 'react-redux';
import muiThemeable from 'material-ui/styles/muiThemeable';
import Loading from '../shared/Loading';
import PokemonDetail from './PokemonDetail';
import { getPokemonDetails } from '../actions';

export class DetailContainer extends Component {

  componentWillMount() {
    const { url } = this.props.location.query;
    this.props.getPokemonDetail(url);
  }

  render() {
    const { isLoading, detailData } = this.props;

    if (isLoading) {
      return <Loading />;
    }

    return (
      <PokemonDetail pokemon={detailData}/>
    );
  }
}

export const mapStateToProps = (state) => {
  const { details } = state;

  return {
    isLoading: details.isLoading,
    detailData: details.results
  };
};

export const mapDispatchToProps = (dispatch) => ({
  getPokemonDetail: (url) => dispatch(getPokemonDetails(url))
});

export default muiThemeable()(connect(mapStateToProps, mapDispatchToProps)(DetailContainer));
