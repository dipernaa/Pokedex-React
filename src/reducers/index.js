import { combineReducers } from 'redux';
import details from './details';
import pokedex from './pokedex';

export default combineReducers({
  details,
  pokedex
});
