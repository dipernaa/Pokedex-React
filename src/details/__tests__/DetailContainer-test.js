import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import mockStore from 'redux-mock-store';
import { DetailContainer, mapStateToProps, mapDispatchToProps } from '../DetailContainer';
import { initialState } from '../../reducers/details';

const mockReduxStore = mockStore({});
const TEST_PROPS = {
  getPokemonDetail: jest.fn(),
  location: {
    query: {
      some: 'value'
    }
  }
};

beforeEach(() => {
  mockReduxStore.clearActions();
});

it('renders a PokemonDetail when not loading', () => {
  const props = {
    ...TEST_PROPS,
    isLoading: false
  };

  const wrapper = shallow(
    <DetailContainer {...props}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders a Loading bar when loading', () => {
  const props = {
    ...TEST_PROPS,
    isLoading: true
  };

  const wrapper = shallow(
    <DetailContainer {...props}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('correctly maps state to props', () => {
  const state = {
    details: initialState
  };

  const props = mapStateToProps(state);
  expect(props).toMatchSnapshot();
});

it('correctly maps dispatch to props', () => {
  const props = mapDispatchToProps(mockReduxStore.dispatch);
  expect(props).toMatchSnapshot();

  const testResponse = { leap: 'rules' };

  fetch.mockResponseSuccess(testResponse);
  return props.getPokemonDetail('http://www.some.url')
    .then(() => {
      expect(mockReduxStore.getActions()).toMatchSnapshot();
    });
});
