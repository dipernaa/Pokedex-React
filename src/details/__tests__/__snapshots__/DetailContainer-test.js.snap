// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly maps dispatch to props 1`] = `
Object {
  "getPokemonDetail": [Function],
}
`;

exports[`correctly maps dispatch to props 2`] = `
Array [
  Object {
    "type": "GET_POKEMON_DETAILS_REQUEST",
  },
  Object {
    "leap": "rules",
    "type": "GET_POKEMON_DETAILS_SUCCESS",
  },
]
`;

exports[`correctly maps state to props 1`] = `
Object {
  "detailData": null,
  "isLoading": false,
}
`;

exports[`renders a Loading bar when loading 1`] = `<Loading />`;

exports[`renders a PokemonDetail when not loading 1`] = `<PokemonDetail />`;
